### ISSUE TITLE
Include a subresource integrity tag on inline `script` tags emitted by the Hoplon compiler

### ISSUE THREAD

**Issue by daveyarwood** [hoplon/hoplon#256]
At Adzerk, I'm working on bringing a Hoplon application up to speed with the security standards codified in [Mozilla Observatory](https://observatory.mozilla.org/).

One major aspect of this (accounting for 50/100 points on the scorecard) is [subresource integrity](https://infosec.mozilla.org/guidelines/web_security#subresource-integrity). The tl;dr is that every inline `script` and `link` element needs to have an `integrity` property, the value of which is a cryptographic hash of the contents of the file being loaded. This allows the browser to verify the integrity of the file being loaded, and refuse to load it if the contents are different from what was expected.

In the case of the Adzerk app I'm working on, the Hoplon compiler writes an HTML page that includes a script tag that loads the compiled JavaScript code, e.g.:

```javascript
<script type="text/javascript" src="index.776e79cbfd7b411ea81aa8f9feeea29c.html.js"></script>
```

In order to have a tight Content Security Policy, I need to add a property like `integrity="sha384-oqVuAfXRKap7fdgcCY5uykM6+R9GqQ8K/uxy9rx7HNQlGYl1kPzQho1wx4JwY8wC"`, where the value is a hash of the contents of that `.html.js` file.

I could do this myself in a post-processing step, but is this something that we would be interested in adding into Hoplon itself?

It seems like it would just be a matter of augmenting [the code here](https://github.com/hoplon/hoplon/blob/1e73edf7c6e56a4c17a0830cb7f3433cdb039c26/src/hoplon/boot_hoplon/compiler.clj#L116-L117). I'd be happy to take a stab at a PR, if this is worth spending time on. It's code that I would need to write anyway if I were to go the post-processing route.

**Comment by flyboarder** [comment]
@daveyarwood I would be happy to welcome a PR for this change. Generally I would say this is a build issue not a hoplon issue, but in the case of boot-hoplon I think we should provide it.

**Comment by daveyarwood** [comment]
After working on this all day yesterday, I realized that one can have a Content Security Policy that includes `script-src 'self'`, meaning that any `script` tags that load a script from the same domain (i.e. `<script src="some-file-hosted-by-me.js"></script>`) are to be trusted. This actually solves the problem I described above in a neater way, so I'll close this issue.

Worth noting (maybe it will help someone who ends up here via search results): I'm running into more issues in the output of the ClojureScript compiler, which includes _inline_ script tags (i.e. `<script>alert("uh-oh!");</script>`). For example:

```javascript
var CLOSURE_UNCOMPILED_DEFINES = {};
var CLOSURE_NO_DEPS = true;
if(typeof goog == "undefined") document.write('<script src="index.9828b13a18fb453fa3cbc64177f178be.html.out/goog/base.js"></script>');
document.write('<script src="index.9828b13a18fb453fa3cbc64177f178be.html.out/goog/deps.js"></script>');
document.write('<script src="index.9828b13a18fb453fa3cbc64177f178be.html.out/cljs_deps.js"></script>');
document.write('<script>if (typeof goog == "undefined") console.warn("ClojureScript could not load :main, did you forget to specify :asset-path?");</script>');
document.write('<script>goog.require("process.env");</script>');
document.write('<script>goog.require("boot.cljs.main3526");</script>');
```

This JS code injects script tags into the page. The first ones that aren't inline (i.e. `<script src="...">`) are actually totally fine if you have a CSP that includes `script-src 'self'`. The last 3, however, are inline script tags and thus need to have an `integrity` property.

I'm actually not sure how the ClojureScript compiler could even fix this on their end. It really does seem like a build issue. The solution I'm working on is to post-process the build in this way:

1. I parse out the contents of each unsafe script tag.
2. I create a base64-encoded SHA256 of each string of content.
3. I include those SHAs in the `script-src` part of my CSP so that the browser knows they can be trusted.
